dask_planner/src/sql/preoptimizer.rs

```diff
@@ -0,0 +1,212 @@
+use std::collections::HashMap;
+
+use datafusion_python::{
+    datafusion::arrow::datatypes::{DataType, TimeUnit},
+    datafusion_common::{Column, DFField, ScalarValue},
+    datafusion_expr::{
+        logical_plan::Filter,
+        utils::from_plan,
+        BinaryExpr,
+        Expr,
+        LogicalPlan,
+        Operator,
+    },
+};
+
+// Sometimes, DataFusion's optimizer will raise an OptimizationException before we even get the
+// chance to correct it anywhere. In these cases, we can still modify the LogicalPlan as an
+// optimizer rule would, however we have to run it independently, separately of DataFusion's
+// optimization framework. Ideally, these "pre-optimization" rules aren't performing any complex
+// logic, but rather "pre-processing" the LogicalPlan for the optimizer. For example, the
+// datetime_coercion preoptimizer rule fixes a bug involving Timestamp-Int operations.
+
+// Helper function for datetime_coercion rule, which returns a vector of columns and literals
+// involved in a (possibly nested) BinaryExpr mathematical expression, and the mathematical
+// BinaryExpr itself
+fn extract_columns_and_literals(expr: &Expr) -> Vec<(Vec<Expr>, Expr)> {
+    let mut result = Vec::new();
+    if let Expr::BinaryExpr(b) = expr {
+        let left = *b.left.clone();
+        let right = *b.right.clone();
+        if let Operator::Plus
+        | Operator::Minus
+        | Operator::Multiply
+        | Operator::Divide
+        | Operator::Modulo = &b.op
```

Comment:
It doesn't really make sense to multiply/divide/modulo a datetime, so should maybe remove these.